### ISSUE TITLE
Mini Support

### ISSUE THREAD

**Issue by ProbablePrime** [Lange/node-elgato-stream-deck#56]
https://www.elgato.com/en/gaming/stream-deck-mini

I'm impulsive so i ordered one

**Comment by ProbablePrime** [comment]
My mini is delivered :D

**Comment by ProbablePrime** [comment]
Ok, I thought it would use the same protocol so i just hacked in a bunch of checks for the number of keys. NOPE different protocol. 

THIS WILL BE FUN!

**Comment by ProbablePrime** [comment]
It looks like the mini writes in a more high fidelity mode. I see Multiple packets for each button. 

i see up to 0x13(19) in the 3rd Byte. We previously called this the page number and we only found 2 now there are 19. I will plug the original in and see if that too has changed.

**Comment by ProbablePrime** [comment]
Regular streamdeck does not do this. I bought a second stream deck for my office last month. I'll capture this one to see if this is a time or purchase thing.

It looks like the streamdeck mini might use a different type of key or logic board.

**Comment by wischi-chr** [comment]
I looked at the preview images on amazon and the icons are much clearer on the mini. Maybe the resolution is better or is it still 72x72 ?

**Comment by wischi-chr** [comment]
@ProbablePrime My Stream Deck Mini arrived today and I looked into that already. Looks like the protocol is the same, but they changed the report length from 8191 to 1024 and thus more "pages" are needed to transfer the BMP file.

**Comment by ProbablePrime** [comment]
I have an email from someone in my inbox on the writing changes you're right that they have longer lengths but there's also some other settings that get changed. Due to the changes I'm going to write a class based structure to handle writing. 
```
MiniWriter extends BaseWriter
RegularWriter extends BaseWriter
```

**Comment by wischi-chr** [comment]
A few things that changed (let me know if you found something else that changed)

- The page number (byte 3) now starts with 0 instead of one
- The keyId order changed (was right to left on SD and is now left to right on SDM)
- The BMP files that are transfered are now 80x80px (but according to elgato systems the LCDs are still 72x72)
- The images are rotated 90° CW (so you need to rotate them CCW)

**Comment by howkj1** [comment]
@ProbablePrime @wischi-chr 
Did you happen to figure out the data needed to get button images / colors working on the mini?
I have success capturing key-press up's and down's with my mini on ubuntu 18.04 but get crashes when I call something like : 
`myStreamDeck.fillColor(2, 254, 0, 0);`
I think its due to page numbers and buffer sizes being wrong?
I'm willing to do some testing but need some ideas!

**Comment by wischi-chr** [comment]
@howkj1 StreamDeck Mini has icon sizes of 80x80 px and the "magic bytes" that precedes the image data is just a regular bmp header. I'm not very good at javascript so I can't diagnose the problems with node-elgato-stream-deck.

But if you need a reference to look things up you can take a look at my HID implementation for [Stream Deck](https://github.com/OpenMacroBoard/StreamDeckSharp/blob/master/src/StreamDeckSharp/Internals/StreamDeckHardwareInfo.cs) and [Stream Deck Mini](https://github.com/OpenMacroBoard/StreamDeckSharp/blob/master/src/StreamDeckSharp/Internals/StreamDeckMiniHardwareInfo.cs). `bmpHeader` and `GeneratePayload` are the things you want (I guess).

**Comment by Julusian** [comment]
In theory, this library now supports the Mini but I do not have one to be able to test it. I would be really grateful if someone is able to test it

**Comment by lexmiln** [comment]
Hey, I've been subscribed to this issue hoping for support for the stream deck mini - I'm doing my best to put the new changes to the test but setting up a sane dev environment on Windows 10 has me going around and around. It looks like I need to install Visual Studio 2015 C++ something something to compile the usb-detection node library from source??? Will update if I make progress (or I'll just try this on a mac instead)

**Comment by lexmiln** [comment]
![image](https://user-images.githubusercontent.com/230271/58680461-71f29100-832d-11e9-82c9-1475b73b7366.png)

Not bad!!!! The images appear rotated and flipped, and appear to be offset and wrapping in a curious way, but this is otherwise working for me.

The photo is from `examples/text-generation.js` - buttons _should_ be cleared when they're unpressed but I was able to get the images to stick if I pressed the button very momentarily. I don't know if that's a bug that's only reproducible on the Mini.

I'mr running Windows 10, node v8.16.0 (the most recent version I could get `node-hid` to build on because of a NewInstance signature issue), with Visual Studio 2017 (including the Visual C++ ATL for x86 and x64 component) and I had to run `npm config set msvs_version 2017` as well, in case that helps some poor future Googler.

I was able to trivially patch `examples/brightness.js` for six buttons and adjusting the brightness is working.

Even with the way-off bitmap rendering, this totally unblocks me for the project I bought the Mini for in the first place. Thanks so much for maintaining the project!

**Comment by Lange** [comment]
Try  https://www.npmjs.com/package/windows-build-tools

**Comment by Julusian** [comment]
@lexmiln Thanks for giving this a test.
Needing vs2015 is annoying, but as usb-detection is only needed for one of the examples it could be left out for testing.

> Not bad!!!! The images appear rotated and flipped, and appear to be offset and wrapping in a curious way, but this is otherwise working for me

Seeing the icons being wrong is really helpful, being able to visualise what it is doing wrong will make it a lot easier to correct. The offset is the slightly most interesting part of that, Is that strip of white at the top of the buttons a reflection or part of the display?

> I was able to get the images to stick if I pressed the button very momentarily. I don't know if that's a bug that's only reproducible on the Mini.

I have noticed the sticky issue in this example with mine, and is something I have been meaning to look into.

> I'mr running Windows 10, node v8.16.0 (the most recent version I could get node-hid to build on because of a NewInstance signature issue),

As for not working on node 10, that will be fixed very soon, I believe we are no longer blocked upstream on that issue.

> Even with the way-off bitmap rendering, this totally unblocks me for the project I bought the Mini for in the first place. Thanks so much for maintaining the project!

I'm glad to hear that, I'll try to fix that rendering in the next couple of days so keep an eye out for the changes :)

**Comment by lexmiln** [comment]
| top left | bottom right |
| --- | ----|
| ![image](https://user-images.githubusercontent.com/230271/58705484-fe747200-8374-11e9-905a-ba1d7b419a8f.png) | ![image](https://user-images.githubusercontent.com/230271/58705496-0a603400-8375-11e9-9689-92baf087a01a.png) |

The image appears to be offset down and right, and wrapping around back to the top (yup, that's not a reflection). But even curiouser, the wrapped part of the image seems to be offset to the right by one pixel fewer. I hope that's possible to make out in the pics above.

I took two to show that the bitmap is nicely underdrawing the button frame in the bottom right (so the image has a nice rounded corner) but not even the wrapped part of the image is using the top-left-most pixels available under the button.

**Comment by howkj1** [comment]
It was mentioned in an earlier thread on the mini that it is possibly using
an 80x80px image buffer instead of the 72x72px per key that the regular
deck ueed. What size icons are you pushing to the mini? That might have
something to do with the wrapping.

**Comment by Julusian** [comment]
The code is padding the images by 4px each side as from what I have read the extra resolution is not visible

I am wondering if the image is being split into chunks in the wrong order. I have an idea to debug this myself, so Ill see how that goes and let you both know what I find

**Comment by howkj1** [comment]
@lexmiln @julusian

https://github.com/Lange/node-elgato-stream-deck/issues/56#issuecomment-412319050

**Comment by Julusian** [comment]
@howkj1 That could have been where I read it before

@lexmiln I have pushed some changes, that should improve on this. Could you retest and let me know how it looks?

**Comment by lexmiln** [comment]
@Julusian With the most recent commit 15ce1f27bb0490206b4973499bce5d0a7b5e3bc6 none of the examples display anything on the SDM. Button presses are still working, as is the brightness control, but the buttons are always blank.

I `git bisect`ed and the regression seems to be in the latest commit:

```
C:\Users\alex\Projects\Deck\node-elgato-stream-deck (HEAD detached at 8916aec -> origin) (elgato-stream-deck@2.1.1)
λ git bisect good
15ce1f27bb0490206b4973499bce5d0a7b5e3bc6 is the first bad commit
commit 15ce1f27bb0490206b4973499bce5d0a7b5e3bc6
Author: Julian Waller <git@julusian.co.uk>
Date:   Fri May 31 18:42:50 2019 +0100

    fix: Refactor fillImage command generation, and some attempted fixes for the mini

:040000 040000 fabd064b5baf0a723408995ecf332b362fa4acf7 a7068fd336a3efcb8bb0e0562c7257d6d352f826 M      src
:100644 100644 f3ceec383fe09135bd5d33a572e71bb474758977 2c27790874c3e7dd3f294e39c1cd35ff0f1e1581 M      tslint.json
```

**Comment by Julusian** [comment]
@lexmiln could you try changing https://github.com/Lange/node-elgato-stream-deck/commit/15ce1f27bb0490206b4973499bce5d0a7b5e3bc6#diff-a8c2776a92bf173b9aac3717418d2de5R124 back to `const firstPart = 0`?

**Comment by lexmiln** [comment]
With pleasure.

| high bracket | low bracket |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/230271/58750490-eaa23c00-8458-11e9-97d1-86674a6115b4.png) | ![image](https://user-images.githubusercontent.com/230271/58750494-f2fa7700-8458-11e9-9a4d-bdedcd390723.png) |

- Nailed it - images are displaying again
- Beautifully centered
- Flipped horizontally! Both the text and octocat's tail are backwards
- It looks like it's not true that the LCDs are still 72x72 - comparing these new pics with the early ones, it's clear that before we were unintentionally underdrawing the rounded corner. But now with the padding, the image is inset, and we're missing out on rendering all the way to the edges of the display.

Amazing work @Julusian, thanks again for your time and effort on this.

**Comment by Julusian** [comment]
that was an easy fix then :)

To confirm the image size, if you add `byteBuffer.fill(120)` to line 29 of util.ts (https://github.com/Lange/node-elgato-stream-deck/blob/master/src/util.ts#L29) that will fill the space outside the picture in grey, so it will be nice and easy to confirm that the extra resolution is visible.

I think the flip should be fixed by setting line 152 of models.ts (https://github.com/Lange/node-elgato-stream-deck/blob/master/src/models.ts#L152) to `return { x: this.IMAGE_SIZE - y - 1, y: x }`

**Comment by lexmiln** [comment]
![image](https://user-images.githubusercontent.com/230271/58750870-04de1900-845d-11e9-8adf-9bb3b857c5fc.png)

Flip fix is good. Grey pixels abound.

Perhaps it is because of the lensing effect that we see especially on button #5 in this photo that it is preferred to leave the border pixels unused? I'm not sure, I can't easily compare to what the official software does.

**Comment by Julusian** [comment]
Perhaps, or maybe they scale everything in their software up from 72 to 80, so that it appears no different to normal users. Either way I think it should be up to the library user to pad however much they want to, so I shall remove it. 

I shall remove the padding logic and push these other fixes shortly.

**Comment by Julusian** [comment]
@lexmiln latest master should fix those issues

**Comment by wischi-chr** [comment]
@lexmiln 

> It looks like it's not true that the LCDs are still 72x72 - comparing these new pics with the early ones, it's clear that before we were unintentionally underdrawing the rounded corner. But now with the padding, the image is inset, and we're missing out on rendering all the way to the edges of the display.

That's true the LCDs are 80x80 pixel on the stream deck mini (here is prove for that: https://www.youtube.com/watch?v=xjthoJevvKw) I talked to elgato about that back in August 2018 and it sounded like the support (they told me that they event talked to developers) wasn't even aware of that fact.

Long story short - the stream deck mini LCDs are 80x80 and you can control all 80x80 pixels, the elgato stream deck software does some corner rounding magic and only supports 72x72 icons.

**Comment by lexmiln** [comment]
@Julusian `master` fixes those issues :+1:

**Comment by Julusian** [comment]
@wischi-chr that is interesting to know. annoying that they were misleading, but not a big deal really.

@lexmiln Thats great to hear. I shall close this issue once I have done a release unless someone reports any other problems :)

**Comment by Julusian** [comment]
Closing as this was done a few months ago, and no further discussion has occured.
If there is anything wrong, please open a new issue
